sql: TestSqlActivityUpdateTopLimitJob failed

sql.TestSqlActivityUpdateTopLimitJob [failed](https://mesolite.cluster.engflow.com/invocations/default/659acc54-a281-4404-994c-6f02fe0cdc83?testReportRun=1&testReportShard=12&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on release-25.4.3-rc @ [54441a9bd0bd3f69647bff398a3135e70eefad69](https://github.com/cockroachdb/cockroach/commits/54441a9bd0bd3f69647bff398a3135e70eefad69):


```
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecprojconst.projJSONFetchValJSONBytesConstOp.Next
        	            	  | 	bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecprojconst/proj_const_right_ops.eg.go:22202
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*vectorTypeEnforcer).Next
        	            	  | 	pkg/sql/colexec/colexecutils/operator.go:152
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecprojconst.projJSONFetchValJSONBytesConstOp.Next
        	            	  | 	bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecprojconst/proj_const_right_ops.eg.go:22202
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*vectorTypeEnforcer).Next
        	            	  | 	pkg/sql/colexec/colexecutils/operator.go:152
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecprojconst.projJSONFetchTextJSONBytesConstOp.Next
        	            	  | 	bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecprojconst/proj_const_right_ops.eg.go:22620
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*vectorTypeEnforcer).Next
        	            	  | 	pkg/sql/colexec/colexecutils/operator.go:152
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase.(*castStringFloatOp).Next
        	            	  | 	bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecbase/cast.eg.go:10206
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*vectorTypeEnforcer).Next
        	            	  | 	pkg/sql/colexec/colexecutils/operator.go:152
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecproj.projMultFloat64Float64Op.Next
        	            	  | 	bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecproj/proj_non_const_ops.eg.go:11860
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*BatchSchemaSubsetEnforcer).Next
        	            	  | 	pkg/sql/colexec/colexecutils/operator.go:219
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*bufferOp).advance
        	            	  | 	pkg/sql/colexec/buffer.go:43
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*caseOp).Next
        	            	  | 	pkg/sql/colexec/case.go:189
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*BatchSchemaSubsetEnforcer).Next
        	            	  | 	pkg/sql/colexec/colexecutils/operator.go:219
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*bufferOp).advance
        	            	  | 	pkg/sql/colexec/buffer.go:43
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*caseOp).Next
        	            	  | 	pkg/sql/colexec/case.go:189
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*BatchSchemaSubsetEnforcer).Next
        	            	  | 	pkg/sql/colexec/colexecutils/operator.go:219
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*bufferOp).advance
        	            	  | 	pkg/sql/colexec/buffer.go:43
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*caseOp).Next
        	            	  | 	pkg/sql/colexec/case.go:189
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase.(*simpleProjectOp).Next
        	            	  | 	pkg/sql/colexec/colexecbase/simple_project.go:113
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
        	            	  | 	pkg/sql/colexec/invariants_checker.go:88
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
        	            	  | 	pkg/sql/colflow/stats.go:120
        	            	Wraps: (9) could not parse "\"NaN\"" as type float
        	            	Wraps: (10) strconv.ParseFloat: parsing "\"NaN\""
        	            	Wraps: (11) invalid syntax
        	            	Error types: (1) *secondary.withSecondaryError (2) *withstack.withStack (3) *errutil.withPrefix (4) *colexecerror.notInternalError (5) *colexecerror.notInternalError (6) *colexecerror.notInternalError (7) *pgerror.withCandidateCode (8) *withstack.withStack (9) *errutil.withPrefix (10) *strconv.NumError (11) *errors.errorString
        	Test:       	TestSqlActivityUpdateTopLimitJob
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestSqlActivityUpdateTopLimitJob2490059537
--- FAIL: TestSqlActivityUpdateTopLimitJob (8.75s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=12</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSqlActivityUpdateTopLimitJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58403

--- a/pkg/sql/appstatspb/app_stats.go
+++ b/pkg/sql/appstatspb/app_stats.go
@@ -122,16 +122,21 @@ func (t *TransactionStatistics) Add(other *TransactionStatistics) {
 	if other.MaxRetries > t.MaxRetries {
 		t.MaxRetries = other.MaxRetries
 	}
-
-	t.IdleLat.Add(other.IdleLat, t.Count, other.Count)
-	t.CommitLat.Add(other.CommitLat, t.Count, other.Count)
-	t.RetryLat.Add(other.RetryLat, t.Count, other.Count)
-	t.ServiceLat.Add(other.ServiceLat, t.Count, other.Count)
-	t.NumRows.Add(other.NumRows, t.Count, other.Count)
-	t.RowsRead.Add(other.RowsRead, t.Count, other.Count)
-	t.BytesRead.Add(other.BytesRead, t.Count, other.Count)
-	t.RowsWritten.Add(other.RowsWritten, t.Count, other.Count)
-	t.KVCPUTimeNanos.Add(other.KVCPUTimeNanos, t.Count, other.Count)
+	transactionStatsCount := t.Count
+	if transactionStatsCount == 0 && other.Count == 0 {
+		// If both are zero, artificially set the receiver's count to one to avoid
+		// division by zero in Add.
+		transactionStatsCount = 1
+	}
+	t.IdleLat.Add(other.IdleLat, transactionStatsCount, other.Count)
+	t.CommitLat.Add(other.CommitLat, transactionStatsCount, other.Count)
+	t.RetryLat.Add(other.RetryLat, transactionStatsCount, other.Count)
+	t.ServiceLat.Add(other.ServiceLat, transactionStatsCount, other.Count)
+	t.NumRows.Add(other.NumRows, transactionStatsCount, other.Count)
+	t.RowsRead.Add(other.RowsRead, transactionStatsCount, other.Count)
+	t.BytesRead.Add(other.BytesRead, transactionStatsCount, other.Count)
+	t.RowsWritten.Add(other.RowsWritten, transactionStatsCount, other.Count)
+	t.KVCPUTimeNanos.Add(other.KVCPUTimeNanos, transactionStatsCount, other.Count)
 
 	t.ExecStats.Add(other.ExecStats)
 
@@ -171,18 +176,24 @@ func (s *StatementStatistics) Add(other *StatementStatistics) {
 	if other.MaxRetries > s.MaxRetries {
 		s.MaxRetries = other.MaxRetries
 	}
+	statementStatsCount := s.Count
+	if statementStatsCount == 0 && other.Count == 0 {
+		// If both are zero, artificially set the receiver's count to one to avoid
+		// division by zero in Add.
+		statementStatsCount = 1
+	}
 	s.SQLType = other.SQLType
-	s.NumRows.Add(other.NumRows, s.Count, other.Count)
-	s.IdleLat.Add(other.IdleLat, s.Count, other.Count)
-	s.ParseLat.Add(other.ParseLat, s.Count, other.Count)
-	s.PlanLat.Add(other.PlanLat, s.Count, other.Count)
-	s.RunLat.Add(other.RunLat, s.Count, other.Count)
-	s.ServiceLat.Add(other.ServiceLat, s.Count, other.Count)
-	s.OverheadLat.Add(other.OverheadLat, s.Count, other.Count)
-	s.BytesRead.Add(other.BytesRead, s.Count, other.Count)
-	s.RowsRead.Add(other.RowsRead, s.Count, other.Count)
-	s.RowsWritten.Add(other.RowsWritten, s.Count, other.Count)
-	s.KVCPUTimeNanos.Add(other.KVCPUTimeNanos, s.Count, other.Count)
+	s.NumRows.Add(other.NumRows, statementStatsCount, other.Count)
+	s.IdleLat.Add(other.IdleLat, statementStatsCount, other.Count)
+	s.ParseLat.Add(other.ParseLat, statementStatsCount, other.Count)
+	s.PlanLat.Add(other.PlanLat, statementStatsCount, other.Count)
+	s.RunLat.Add(other.RunLat, statementStatsCount, other.Count)
+	s.ServiceLat.Add(other.ServiceLat, statementStatsCount, other.Count)
+	s.OverheadLat.Add(other.OverheadLat, statementStatsCount, other.Count)
+	s.BytesRead.Add(other.BytesRead, statementStatsCount, other.Count)
+	s.RowsRead.Add(other.RowsRead, statementStatsCount, other.Count)
+	s.RowsWritten.Add(other.RowsWritten, statementStatsCount, other.Count)
+	s.KVCPUTimeNanos.Add(other.KVCPUTimeNanos, statementStatsCount, other.Count)
 	s.Nodes = util.CombineUnique(s.Nodes, other.Nodes)
 	s.KVNodeIDs = util.CombineUnique(s.KVNodeIDs, other.KVNodeIDs)
 	s.Regions = util.CombineUnique(s.Regions, other.Regions)

